### ISSUE TITLE
Strange things with LinBASEX

### ISSUE THREAD

**Issue by MikhailRyazanov** [PyAbel/PyAbel#356]
In continuing #355, I was investigating strange things in the `Transform` class, and some of them led me to even stranger things related to LinBASEX.

# Quadrants #
This part:
https://github.com/PyAbel/PyAbel/blob/abe87261730564d76528484cb77a71e79b0b4a2b/abel/transform.py#L580-L585
looks very strange and has an obviously wrong syntax. However, it doesn't cause errors because is _never executed_ ... because in this code:
https://github.com/PyAbel/PyAbel/blob/abe87261730564d76528484cb77a71e79b0b4a2b/abel/transform.py#L504-L505
`_symmetry_axis` is never `None` (passing `symmetry_axis=None` in fact [sets](https://github.com/PyAbel/PyAbel/blob/abe87261730564d76528484cb77a71e79b0b4a2b/abel/transform.py#L470-L475) `symmetry_axis = [None]`, notice the list).

Returning to the “speed/anisotropy”, I guess, the intent was something like
```python
            # speed and anisotropy
            self.Beta = [(Beta0[0] + Beta1[0] + Beta2[0] + Beta3[0]) / 4,
                         (Beta0[1] + Beta1[1] + Beta2[1] + Beta3[1]) / 4]
```
(`self.linbasex_angular_integration` and `self.linbasex_anisotropy_parameter` are in fact not documented — this is how I've strated investigating this). Another problem is that `self.radial` and `self.projection` are not assigned in his case, although they must be, according to the documentation (the question is, however, what to do with them: also average over quadrants or return quadrant-wise 4-tuples?).

The latter problem revealed to an even more strange thing...

# linbasex_transform_full() return even-sized image #

The output of the full-image LinBASEX transform is even-sized, contrary to everything:
```python
import numpy as np
import abel
im = np.ones((9, 9))
tr, _, _, _ = abel.linbasex.linbasex_transform_full(im)
print(im.shape)  # (9, 9)
print(tr.shape)  # (10, 10)
tr = abel.Transform(im, method='linbasex').transform
print(tr.shape)  # (10, 10)
```
You can also plot the resulting image and see that it's shifted as if 1-based indexing was implied somewhere. And the extremely strange thing is that the unit test actually asserts that returning a 22 × 22 output for a 21 × 21 input is the way to go:
https://github.com/PyAbel/PyAbel/blob/abe87261730564d76528484cb77a71e79b0b4a2b/abel/tests/test_linbasex.py#L15-L21
I don't think that this is correct. Any other opinions? And what causes the error?

(`linbasex_transform()` returns an odd-sized image but obviously distorted due to the offset in the full-image transform, which it uses internally.)

# linbasex_transform_full() completely ignores return_Beta #
Well, [see the code](https://github.com/PyAbel/PyAbel/blob/abe87261730564d76528484cb77a71e79b0b4a2b/abel/linbasex.py#L174-L207), it always returns everything. I think, it should comply with its own documentation and instead of directly returning the `_linbasex_transform_with_basis()` results, discard `radial, Beta, projections` when not asked for. A potential problem, however, is that the default parameter is actually `return_Beta=False`, so correcting this behavior will break compatibility. Should we then change it to `return_Beta=True`? Or hope that nobody used this function directly? Or that they will be willing to repair their code without complains, since they didn't care to report this issue? :–)

# int_beta() has unused parameter radial_step #
Again, [see the code](https://github.com/PyAbel/PyAbel/blob/abe87261730564d76528484cb77a71e79b0b4a2b/abel/linbasex.py#L321-L359). Was it supposed to do something? Or should be removed?

**Comment by stggh** [comment]
This is my sloppy coding, in a poor attempt to integrate Thomas Gerber's `Lin_Basex_0.7` python notebook into `PyAbel`, issue #167. 

From memory, `linbasex` operates on even sized images, and returns the full image transform, along with the anisotropy parameters. Some fudging was done to make it emulate the return parameters of the other transform methods. Clearly, I failed. However. I am puzzled as I have made use of the `linbasex` method to obtain the anisotropy parameter as a function of radius, before you made available this parameter in the other basex methods.

I will take a look at the code over the next few days.

**Comment by MikhailRyazanov** [comment]
I guess, it wasn't so much sloppiness but that PyAbel itself was in its early stage and Lin-BASEX was different from everything else...

I didn't find anything specific about that `linbasex` requires an even image size, except mentioning that its output is for some reason even-sized, but that the assumed origin is actually pixel-centered (see [here]((https://github.com/PyAbel/PyAbel/issues/167#issuecomment-214668091)) and several following comments), which would imply an odd image size. Visually, this seems to be correct.

In the meantime, I've also looked at the code, and to me the problem is here:
https://github.com/PyAbel/PyAbel/blob/abe87261730564d76528484cb77a71e79b0b4a2b/abel/linbasex.py#L299
and, correspondingly, here:
https://github.com/PyAbel/PyAbel/blob/abe87261730564d76528484cb77a71e79b0b4a2b/abel/linbasex.py#L312-L314
If these are replaced by
```python
    Slice_3D = np.zeros((pol, 2 * NP - 1, 2 * NP - 1))
```
(removing the extra row and column) and
```python
    for i in range(pol):
        Slice_3D[i] = np.fromfunction(lambda k, l: _SL(i, k - NP + 1, l - NP + 1,
                       Beta_convol, index, legendre_orders), Slice_3D[i].shape)
```
(shifting both indices by −1), then the results become odd-sized and correctly centered.
The latter code fragment in fact doesn't need `fromfunction` and can be replaced by passing corresponding arrays directly to `_SL`:
```python
    col = np.arange(-NP + 1, NP)
    row = col[:, None]
    for i in range(pol):
        Slice_3D[i] = _SL(i, row, col, Beta_convol, index, legendre_orders)
```
which looks simpler and also works slightly faster.

I haven't tested everything, but at the first glance, the output image looks correct. (But I had to also modify `example_linbasex.py`, for example, because it relied on the input image being even-sized...)

**Comment by MikhailRyazanov** [comment]
> Some fudging was done to make it emulate the return parameters of the other transform methods. Clearly, I failed. However. I am puzzled as I have made use of the linbasex method to obtain the anisotropy parameter as a function of radius, before you made available this parameter in the other basex methods.

Well, `linbasex_transform()` and `Transform()` with `method='linbasex'` do this correctly. The only problem is with `linbasex_transform_full()`, which has a parameter `return_Beta=False` but in fact returns `inv_IM, radial, Beta, projections` by default and regardless of the `return_Beta` value. I see 3 possible solutions:
1. Remove the `return_Beta` parameter and always return everything. This will cause some incompatibility with the code that was passing `return_Beta=True`, if such code exists anywhere.
2. Change the default to `return_Beta=True` and modify the function to actually return just the image when `return_Beta=False` is passed explicitly. This will make the behavior reasonable and compatible with default parameters but by default not consistent with `linbasex_transform()` (and `Transform()`, although I also think that `Transform()` with `method='linbasex'` should always save at least `Beta` and `radial` because this costs nothing and is kind of expected from the Lin-BASEX method).
3. Modify the function to implement the `return_Beta` logic without changing the default `return_Beta=False`. This will make it consistent internally and with everything else but will break the user code that used `linbasex_transform_full()` with default parameters.

**Comment by stggh** [comment]
@MikhailRyazanov - thanks for you investigations of this. 

My apologies for the delay in my reply, I am tied up with a home renovation, and some travel planning (Iceland in October). 

The transform image off-set has always puzzled me, and not having your mind for these matters, the issue has elluded me.
Your suggested changes work well.

For `linbasex_transform()` the goal was to have the same return parameters as the other transform methods, but I think that may have changed with `rbasex`. Item 2. was my original intention, always save `Beta` and `radial`, and only return these parameter from `linbasex_transform()` if `return_Beta=True`.

**Comment by MikhailRyazanov** [comment]
Hm, I forgot how it was with rBasex, but now, looking through the PR #270 review, I've found this (GitHub can't link to “collapsed” items, so I'm showing a screenshot):
![rBasex-review](https://user-images.githubusercontent.com/12854133/184689276-b84eda7e-8e16-4b83-a028-ca24befb6fcd.png)
That is, although @DanHickstein has appropriately noticed that `rbasex_transform()` behaves differently from all other methods, there was apparently a consensus that this is not a problem, and perhaps `linbasex` should do the same...

So I would propose that, for compatibility, we leave `linbasex_transform()` as is (except documenting more clearly that it transforms the _upper_ right quadrant, not just any “right side of an image”) but for `linbasex_transform_full()` deprecate the `return_Beta` parameter, conserving the current behavior (always return all the results). And then also modify `Transform()` accordingly to always save `Beta` and `radial` (and ignore `return_Beta` in `transform_options`).

On the other hand, regarding `Transform()` and its imaginary use of `linbasex_transform()` in `_abel_transform_image_by_quadrant()`, I'm now more inclined to remove this completely instead of trying to repair it, taking into account that:
1. It apparently never worked anyway (git blame [shows](https://github.com/PyAbel/PyAbel/commit/a322afe4126261d9834e3f7a53106bedd83e8b41) that the code was broken since at least 2016, before the first [release](https://github.com/PyAbel/PyAbel/releases/tag/v0.7.5b) with linbasex).
2. Lin-BASEX is intended for images with a very specific structure. Transforming them by quadrants could be useful only for checking discrepancies between them, but then the distributions must also be returned per quadrant, which will break the current default behavior (with `symmetry_axis=None`). In addition, the transform with odd Legendre orders by quadrants must be done differently and carefully...
3. Lin-BASEX is already the slowest among practically interesting methods, and doing the transform by quadrants will make it additionally 4 times slower with the default options (`symmetry_axis=None`).

(I've started to prepare the [changes](https://github.com/MikhailRyazanov/PyAbel/tree/linbasex) so that the actual code can be seen/tried before the PR. I'll be adding the rest as we decide what exactly to do.)

**Comment by stggh** [comment]
> prepare the changes

`linbasex` is in good hands!

**Comment by DanHickstein** [comment]
This all sounds good to me! Great work Mikhail!

**Comment by MikhailRyazanov** [comment]
OK, I've removed `return_Beta` from `linbasex_transform_full()` and made sure that `Transform()` always saves `radial`, `Beta` and `projection` (and doesn't try to use `linbasex` by quadrants).

The remaining question is: what `radial_step` in `int_beta()` was supposed to do? Is everything already done by `radial_step` in `linbasex_transform_full()`? Or to get correct results from `Beta`, the same `radial_step` is needed inside `int_beta()` but was neglected?

**Comment by stggh** [comment]
It looks as though `int_Beta()` is not used anywhere.  The code was part of the notebook dump. Kept in case it was important.
In the original notebook code it was listed as an example.
```python
# ## Integrate beta over a range of Newton spheres (example)

# In[59]:

#Define a set of ranges. As many as you need.

regions= range(37//inc,40//inc),range(69//inc,72//inc),range(89//inc,92//inc),range(133//inc,136//inc)

Beta_n=np.zeros(Beta.shape)     #Define new array for normalized beta's, independent of Beat_norm defined above.
max_counts=max(Beta[0,:])       #Normalized to Newton sphere with maximal counts.

threshold=0.00                   #set threshold for normalisation of higher orders, 0.0 ... 1.0. (As above).

Beta_n[0] =Beta[0]/max_counts
for i in range(1,pol):
    Beta_n[i]=np.where(Beta[0]/max_counts>threshold,Beta[i,:]/Beta[0,:],0)
 
Beta_int=np.zeros((pol,len(regions)))     #Define arrays for results
legend=np.zeros((2,len(regions)),dtype=int)

ro=2               #chose number of digits to print

print ("Region, Integrated Beta")
print
print ('{0:5} {1:5} {2:2} {3:5} {4:5}'.format('|from|','|to|','| β'+ str(un[0]), '| β'+str(un[1]),'|  ...' ))
for j in range(len(regions)):
    legend[0,j]=min(regions[j])
    legend[1,j]=max(regions[j])
    for i in range(pol):
        Beta_int[i,j]=sum(Beta_n[i,regions[j]])/len(regions[j])   
    print ('{0:5} {1:5} {2:100}'.format(legend[0,j],legend[1,j], str(Beta_int[:,j].round(ro))))
```

**Comment by MikhailRyazanov** [comment]
Hm, from `... range(37//inc,40//inc) ...` it seems that `radial_step` must be used at least to determine correct `Beta` indices from the radii given in `regions`.

I'm, however, confused with the overall behavior of `int_beta()`:
```python
import numpy as np
import matplotlib.pyplot as plt
import abel
im = abel.tools.analytical.SampleImage(name='O-', sigma=5)
tr, rad, bet, pr = abel.linbasex.linbasex_transform_full(im.abel)
regions = [(98, 106), (118, 126)]
int_bet = abel.linbasex.int_beta(bet, regions=regions)
print(int_bet)
# output:
# [[ 0.45126457  0.81402086]
#  [-0.32132272 -0.1843801 ]]
for n, b in enumerate(bet):
    c = 'C' + str(n)
    plt.plot(rad, b, c=c, label='Beta[{}]'.format(n))
    for reg, ib in zip(regions, int_bet[n]):
        plt.plot(reg, [ib, ib], c=c, lw=2)
plt.legend()
plt.tight_layout()
plt.show()
```
![int_beta](https://user-images.githubusercontent.com/12854133/185207461-089256ff-4b08-4c21-9617-560d203c0e1b.png)
For `Beta[0]` it seems to _average_ instead of _integrate_. For `Beta[1]` the results look completely meaningless.

And a usability remark: `regions` is a required argument, so giving it a default value, particularly `None`, which immediately causes an error, is a strange decision.

I also don't see this function used in any of our code. So maybe it would be better to remove it completely?

**Comment by MikhailRyazanov** [comment]
By the way, I've noticed that, for example, passing `radial_step=2` to `linbasex_transform_full()` makes the transformed image twice smaller (in both dimensions) and 8 times more intense. This is not documented anywhere and is not really something obviously expected. Especially since `radial` covers the whole radial range of the input image (with 2-pixel steps, which is expected).

**Comment by MikhailRyazanov** [comment]
For `radial_step` > 1 in `linbasex_transform_full()`, it was easy to make the output correctly sized, so I've already made the changes. This has also solved the problem with non-zero `clip` mangling the output image (for default `radial_step=1` as well):
```python
im = abel.tools.analytical.SampleImage(name='Gerber')
tr, rad, bet, pr = abel.linbasex.linbasex_transform_full(im.abel, proj_angles=[0, np.pi/4, np.pi/2],
                                                         legendre_orders=[0, 2, 4], clip=...)
```
**before:** (all the radii are smaller by `clip` pixels, and the image size is also reduced correspondingly)
![before](https://user-images.githubusercontent.com/12854133/185262300-4d1594ca-6cc1-412a-97d4-9dc9dd3b89ca.png)
**after:**
![after](https://user-images.githubusercontent.com/12854133/185262360-183d59b3-242c-40bc-9fae-1d1965fc52a5.png)
(`radial` and `Beta` were not affected; their first `clip` elements are indeed “clipped” as documented.)

**Comment by MikhailRyazanov** [comment]
@stggh, I think, I understand what is going on. The notebook code was apparently intended to operate on “raw” `Beta`, which are _intensities_. So it was “normalizing” `Beta[1:]` by `Beta[0]` to convert them  to _anisotropy parameters_ and then averaging (not “integrating”) over each region. The problem with `int_beta()` is that `linbasex_transform_full()` outputs already “normalizing” `Beta`, and then `int_beta()` “normalizes” it dividing by `Beta[0]` again, leading to completely meaningless results.

So I've tried to make a new function `mean_beta()` to do this properly (see 051750f). The results look fine:
```python
import numpy as np
import matplotlib.pyplot as plt
import abel

im = abel.tools.analytical.SampleImage(n=513, name='Gerber')
recon, radial, beta, proj = abel.linbasex.linbasex_transform_full(
    im.abel,
    proj_angles=[0, np.pi/4, np.pi/2],
    legendre_orders=[0, 2, 4],
    threshold=0.01
)

regions = [(34, 41), (66, 73), (86, 93), (130, 146), (192, 199), (226, 233)]
beta_mean = abel.linbasex.mean_beta(radial, beta, regions)

plt.figure(dpi=200)
for n, b in enumerate(beta):
    c = 'C' + str(n)
    plt.plot(radial, b, c=c, lw=0.5, label='Beta[{}]'.format(n))
    if n:
        for reg, ib in zip(regions, beta_mean[n]):
            plt.plot(reg, [ib, ib], c=c)
plt.xlim(0, 256)
plt.legend()
plt.tight_layout()
plt.show()
```
![mean_beta](https://user-images.githubusercontent.com/12854133/185770948-dbf7bbd6-3b50-4f62-93a6-51f9a959d888.png)
But for `Beta[0]` it actually does integration because “average intensity in a region” is not very useful — correct me if I'm wrong.

**Comment by stggh** [comment]
Hi @MikhailRyazanov - I think it would be rare occurrence when you are wrong. 

Given that `Beta[1]` is already correctly scaled, to provide the anisotropy parameter, the product `Beta[1]*Beta[0]` may be processed (e.g. smoothed) as a normal data set. There is no need for any additional peak specific intensity weighting. 

Sorry, if I have missed the point of this function.

**Comment by MikhailRyazanov** [comment]
> the product `Beta[1]*Beta[0]` may be processed (e.g. smoothed) as a normal data set

This is exactly what the new function does. Calculating the mean anisotropy as (∫β(r)⋅I(r) dr)/(∫I(r) dr) _is_ the intensity-weighted β averaging. The function is an analog of [`abel.tools.vmi.radial_integration()`](https://pyabel.readthedocs.io/en/latest/abel.html#abel.tools.vmi.radial_integration), except that it works for anisotropies of any order but can't estimate their errors (lin-BASEX doesn't give such information) and doesn't return the intensity vs angle distributions (they are analytical in lin-BASEX).

**Comment by stggh** [comment]
∫β(r)⋅I(r) dr/∫I(r) dr  is correct.  I refer you to the first statement of my comment [above](https://github.com/PyAbel/PyAbel/issues/356#issuecomment-1221462426).

**Comment by MikhailRyazanov** [comment]
OK, everything is addressed in #357, so this can be closed.
